src/IO/S3/copyDataToS3.cpp

```diff
@@ -0,0 +1,723 @@
+            }
+
+            request.SetMultipartUpload(multipart_upload);
+
+            size_t max_retry = std::max(max_unexpected_write_error_retries, 1UL);
+            for (size_t i = 0; i < max_retry; ++i)
+            {
+                ProfileEvents::increment(ProfileEvents::S3CompleteMultipartUpload);
+
+                auto outcome = client_ptr->CompleteMultipartUpload(request);
+
+                if (outcome.IsSuccess())
+                {
+                    LOG_TRACE(log, "Multipart upload has completed. Bucket: {}, Key: {}, Upload_id: {}, Parts: {}", dest_bucket, dest_key, multipart_upload_id, part_tags.size());
+                    break;
+                }
+                else if (outcome.GetError().GetErrorType() == Aws::S3::S3Errors::NO_SUCH_KEY)
+                {
+                    /// For unknown reason, at least MinIO can respond with NO_SUCH_KEY for put requests
+                    /// BTW, NO_SUCH_UPLOAD is expected error and we shouldn't retry it
+                    LOG_INFO(log, "Multipart upload failed with NO_SUCH_KEY error for Bucket: {}, Key: {}, Upload_id: {}, Parts: {}, will retry", dest_bucket, dest_key, multipart_upload_id, part_tags.size());
+                    /// will retry
+                }
+                else
+                {
+                    throw S3Exception(
+                        outcome.GetError().GetErrorType(),
+                        "Message: {}, Key: {}, Bucket: {}, Tags: {}",
+                        outcome.GetError().GetMessage(), dest_key, dest_bucket, fmt::join(part_tags.begin(), part_tags.end(), " "));
+                }
+            }
+        }
+
+        void abortMultipartUpload()
+        {
+            LOG_TRACE(log, "Aborting multipart upload. Bucket: {}, Key: {}, Upload_id: {}", dest_bucket, dest_key, multipart_upload_id);
+            Aws::S3::Model::AbortMultipartUploadRequest abort_request;
+            abort_request.SetBucket(dest_bucket);
+            abort_request.SetKey(dest_key);
+            abort_request.SetUploadId(multipart_upload_id);
+            client_ptr->AbortMultipartUpload(abort_request);
+            multipart_upload_aborted = true;
+        }
+
+        void checkObjectAfterUpload()
+        {
+            LOG_TRACE(log, "Checking object {} exists after upload", dest_key);
+            S3::checkObjectExists(*client_ptr, dest_bucket, dest_key, {}, {}, "Immediately after upload");
+            LOG_TRACE(log, "Object {} exists after upload", dest_key);
+        }
+
+        void performMultipartUpload(size_t start_offset, size_t size)
+        {
+            calculatePartSize(size);
+            createMultipartUpload();
+
+            size_t position = start_offset;
+            size_t end_position = start_offset + size;
+
+            for (size_t part_number = 1; position < end_position; ++part_number)
+            {
+                if (multipart_upload_aborted)
```

Comment:
Isn't it better to check this right before processUploadPartRequest? In order to stop work as fast as possible if non retryable error occurs. 
My concern is that the code have to wait all task which has been scheduled. If 1000 task were scheduled and only 100 of then is running and finishing with connection timeout. It would take 10 x timeout to finish the upload with an error.

Reply:
It seems like the queue size is pretty big
>         auto queue_size = config.getUInt(".threadpool_writer_queue_size", 1000000);
and all task are about to be pushed to the schedule's queue. In case if first task fail no need to run the rest.

Reply:
> Isn't it better to also check multipart_upload_aborted right before processUploadPartRequest?

Done, I changed the code.